pkg/codegen/docs/gen.go

```diff
@@ -41,53 +41,77 @@ import (
 	"github.com/pulumi/pulumi/pkg/util/contract"
 )
 
-var supportedLanguages = []string{"csharp", "go", "nodejs", "python"}
+var (
+	supportedLanguages = []string{"csharp", "go", "nodejs", "python"}
+	templates          *template.Template
+	packagedTemplates  map[string][]byte
+	docHelpers         map[string]codegen.DocLanguageHelper
+)
 
-var templates *template.Template
-var packagedTemplates map[string][]byte
+func init() {
+	docHelpers = make(map[string]codegen.DocLanguageHelper)
+	for _, lang := range supportedLanguages {
+		switch lang {
+		case "csharp":
+			docHelpers[lang] = dotnet.DocLanguageHelper{}
+		case "go":
+			docHelpers[lang] = go_gen.DocLanguageHelper{}
+		case "nodejs":
+			docHelpers[lang] = nodejs.DocLanguageHelper{}
+		case "python":
+			docHelpers[lang] = python.DocLanguageHelper{}
+		}
+	}
+}
 
-// Header represents the header of each resource markdown file.
-type Header struct {
+// header represents the header of each resource markdown file.
+type header struct {
 	Title string
 }
 
 type exampleUsage struct {
+	// Heading is the title of an example.
 	Heading string
 	Code    string
 }
 
-// Property represents an input or an output property.
-type Property struct {
+// property represents an input or an output property.
+type property struct {
 	Name               string
 	Comment            string
-	Type               PropertyType
+	Type               propertyType
 	DeprecationMessage string
 
 	IsRequired bool
-	// IsInput is a flag to indicate if a property is an input
-	// property.
-	IsInput bool
+	IsInput    bool
 }
 
-// DocNestedType represents a complex type.
-type DocNestedType struct {
-	Name        string
-	APIDocLinks map[string]string
-	Properties  map[string][]Property
+// apiTypeDocLinks represents the links for a type's input and output API doc.
+type apiTypeDocLinks struct {
+	InputType  string
+	OutputType string
 }
 
-// PropertyType represents the type of a property.
-type PropertyType struct {
+// docNestedType represents a complex type.
+type docNestedType struct {
+	Name         string
+	IsOutputType bool
+	APIDocLinks  map[string]apiTypeDocLinks
+	Properties   map[string][]property
```

Comment:
One general thing I totally forgot to mention. I believe this is still the case, but @pgavlin, please correct me if I am wrong or if we're planning on changing this for 2.0... For nested types only, and for Python only, the lang-specific names of the properties should be `camelCase` rather than `snake_case` like everywhere else in Python.

Reply:
Ohh. Do let me know. Though, I think I'll write-up an issue for it to be fixed in a follow-up PR rather than in this one.